torchrl/data/replay_buffers/checkpointers.py

```diff
@@ -0,0 +1,346 @@
+        is_pytree = metadata["is_pytree"]
+        _len = metadata["len"]
+        if is_pytree:
+            if self._load_hooks:
+                raise RuntimeError(
+                    "Loading hooks are not compatible with PyTree storages."
+                )
+            path = Path(path)
+            for local_path, md in metadata["metadata"].items():
+                # load tensor
+                local_path_dot = local_path.replace(".", "/")
+                total_tensor_path = path / (local_path_dot + ".memmap")
+                shape = torch.Size(md["shape"])
+                dtype = _STRDTYPE2DTYPE[md["dtype"]]
+                tensor = MemoryMappedTensor.from_filename(
+                    filename=total_tensor_path, shape=shape, dtype=dtype
+                )
+                # split path
+                local_path = local_path.split(".")
+                # replace potential dots
+                local_path = [_path.replace("_<dot>_", ".") for _path in local_path]
+                if storage.initialized:
+                    # copy in-place
+                    _storage_tensor = storage._storage
+                    # in this case there is a single tensor, so we skip
+                    if local_path != ["_-single-tensor-_"]:
+                        for _path in local_path:
+                            if _path.isdigit():
+                                _path_attempt = int(_path)
+                                try:
+                                    _storage_tensor = _storage_tensor[_path_attempt]
+                                    continue
+                                except IndexError:
+                                    pass
+                            _storage_tensor = _storage_tensor[_path]
+                    _storage_tensor.copy_(tensor)
+                else:
+                    raise RuntimeError(
+                        "Cannot fill a non-initialized pytree-based TensorStorage."
+                    )
+        else:
+            _storage = TensorDict.load_memmap(path)
+            if storage.initialized:
+                dest = storage._storage
+            else:
+                # TODO: This could load the RAM a lot, maybe try to catch this within the hook and use memmap instead
+                dest = None
+            for hook in self._load_hooks:
+                _storage = hook(_storage, out=dest)
+            if not storage.initialized:
+                # this should not be reached if is_pytree=True
+                storage._init(_storage[0])
+                storage._storage.update_(_storage)
+            else:
+                storage._storage.copy_(_storage)
+        storage._len = _len
+
+
+class FlatStorageCheckpointer(TensorStorageCheckpointer):
+    """Saves the storage in a compact form, saving space on the TED format.
+
+    This class explicitly assumes and does NOT check that:
+
+      - done states (including terminated and truncated) at the root are always False;
```

Comment:
What is meant by `root` here, exactly? 🤔

Reply:
root is conventionally what is not in "next"